### DISCUSSION TITLE
MotionManager usage

### DISCUSSION THREAD

**Discussion by c68r69s** [Yaskawa-Global/YMConnect#11] [Q&A] [answered]
Hello,

I am trying to use the library with a YRC1000, however I seem to misunderstand the way the MotionManager
is meant to be used. I should add that I'm new to controlling robots.

So using the library I can connect just fine, and every api i tried apart from the MotionManager seems to work
like expected. The parameters according to https://developer.motoman.com/en/YMConnect/RobotControllerSetup are set,
and the MotoPlus function is set to "used".

My initial understanding was that I could start a "motion mode" via the `MotionManager->MotionStart()` call.
Then I expected it to move through points that were already added, and that I could add points even after that
until I call `MotionStop()` (excluding any error cases etc.).

So with the test code below I tried this with 2 points, and the robot was supposed to move between those 2.
However, at most I can get 1 single movement.

If I have the MotionStart after adding the points (like it is in the sample below), then I get one movement that
works, and then the robot stops. No error, neither on the robot nor via the library, but no movement either.
This does not change if I comment in the "extra stop".

If I remove the "Starting motion (2)", and adding the "Starting motion (1)" at the beginning, no movement ever happens.

Turning on the servos, reading the position and status all work. Also, if there is a motion I do get the "waiting..." messages.

I'd appreciate any clues

```c++
#include "lib/YMConnect.h"

#include <chrono>
#include <exception>
#include <iostream>
#include <thread>

int main(int argc, char *argv[]) {
  const auto check = [](const StatusInfo& info, std::string msg) {
    if (info.IsOk())
      return;

    throw std::runtime_error(msg + ":" + info.ToString());
  };

  StatusInfo status;
  auto con = YMConnect::OpenConnection("172.30.10.10", status);
  check(status, "Connecting");

  check(con->MotionManager->MotionStop(), "Stop motion mode");
  check(con->MotionManager->ClearAllTrajectory(), "Clearing trajectory");
  check(con->ControlCommands->SetServos(SignalStatus::ON), "Setting servo on");

  // using this was the initial idea, but no movement ever happens
  // check(con->MotionManager->MotionStart(), "Starting motion (1)");

  const MotionAccelDecel acc_dcc{
      .accelRatio = 100.0,
      .decelRatio = 100.0,
  };

  const int tool = 0;

  PositionData targetA;
  targetA.coordinateType = CoordinateType::BaseCoordinate;
  targetA.toolNumber = tool;

  targetA.figure = Figure{
      .frontOrBack = FrontOrBack::Front,
      .upperOrLower = UpperOrLower::Uppper,
      .flipOrNoFlip = FlipOrNoFlip::NoFlip,
      .axisAngles{
          AxisAngle::LessThanOrEqual180,
          AxisAngle::LessThanOrEqual180,
          AxisAngle::LessThanOrEqual180,
          AxisAngle::LessThanOrEqual180,
          AxisAngle::LessThanOrEqual180,
          AxisAngle::LessThanOrEqual180,
      },
  };
  targetA.axisData = {
      350.0, 10.0, 500.0, 
      0.0, 0.0, 0.0,
  };

  PositionData targetB;
  targetB.coordinateType = CoordinateType::BaseCoordinate;
  targetB.toolNumber = tool;

  targetB.figure = Figure{
      .frontOrBack = FrontOrBack::Front,
      .upperOrLower = UpperOrLower::Uppper,
      .flipOrNoFlip = FlipOrNoFlip::NoFlip,
      .axisAngles{
          AxisAngle::LessThanOrEqual180,
          AxisAngle::LessThanOrEqual180,
          AxisAngle::LessThanOrEqual180,
          AxisAngle::LessThanOrEqual180,
          AxisAngle::LessThanOrEqual180,
          AxisAngle::LessThanOrEqual180,
      },
  };
  targetB.axisData = {
      350.0, 10.0, 600.0, 
      0.0, 0.0, 0.0,
  };

  PositionData robot_pos;
  ControllerStateData state;
  bool a = true;
  while (true) {
    check(con->ControlGroup->ReadPositionData(ControlGroupId::R1,
                                              CoordinateType::BaseCoordinate, 0,
                                              tool, robot_pos),
          "Read robot pos");

    check(con->Status->ReadState(state), "Reading state");

    if (state.isRunning) {
      std::this_thread::sleep_for(std::chrono::milliseconds(100));
      std::cout << "waiting..." << std::endl;
      continue;
    }

    const auto& target = a ? targetA : targetB;
    a = !a;

    // I tried this together with the MotionStart below, but no change
    // check(con->MotionManager->MotionStop(), "extra stop");

    std::cout << "Sending movement:\n" << target << std::endl;
    LinearMotion motion(ControlGroupId::R1, target, 100, acc_dcc);
    check(con->MotionManager->AddPointToTrajectory(motion), "Add point");

    // works only the first time
    check(con->MotionManager->MotionStart(), "Starting motion (2)");
    std::this_thread::sleep_for(std::chrono::milliseconds(100));
  }
}

```

**Comment by aseamans** [comment]
Hello,

There was a bug that I found in the MotionManager. I fixed it and added a few features in the latest release. Try using this new release and let me know if it worked. 
Also, The motion manager automatically clears the trajectory upon the call to open connection so you don't need to do it at the start. 
There is a new feature that logs errors in motion. This can be found in the same folder as the location of the DLL. If the new update doesn't work, please post this file so I can take a look.

Thank you.

**Comment by c68r69s** [comment]
Hi,

Thanks for the quick reply.

I have tested the new version, and initially it behaved the same for me as the old one. But i could not find the debug log
you mentioned. I did link against the YMConnect library (without the _D), so i tried the YMConnect_D thinking this will probably
be the debug version and might produce the debug log.
While I still cannot see any debug log, with the YMConnect_D library the test code just works.

Do you have any idea what might cause this? I am using the linux version if that matters.

Thanks

**Comment by ted-miller** [comment]
Looking at the code, your initial idea (`Starting motion (1)`) is the correct approach. 

I was just now able to run the robot with this code. It is slightly modified from your original post.

This just rotates the joints R, B, and T back and forth.

Please also note that the sleep inside of `if (state.isRunning)` is not required. You can queue up multiple points during motion to ensure a smooth blended trajectory.

------------------

Code:

```
#include "YMConnect.h"

#include <chrono>
#include <exception>
#include <iostream>
#include <thread>

int main(int argc, char* argv[]) {
    const auto check = [](const StatusInfo& info, std::string msg) {
        if (info.IsOk())
            return;

        throw std::runtime_error(msg + ":" + info.ToString());
        };

    StatusInfo status;
    auto con = YMConnect::OpenConnection("192.168.1.31", status);
    check(status, "Connecting");

    check(con->MotionManager->MotionStop(), "Stop motion mode");
    check(con->MotionManager->ClearAllTrajectory(), "Clearing trajectory");
    check(con->ControlCommands->SetServos(SignalStatus::ON), "Setting servo on");

    // using this was the initial idea, but no movement ever happens
    check(con->MotionManager->MotionStart(), "Starting motion (1)");

    MotionAccelDecel acc_dcc;
    acc_dcc.accelRatio = 100;
    acc_dcc.decelRatio = 100;

    const int tool = 0;

    PositionData targetA;
    targetA.coordinateType = CoordinateType::Pulse;
    targetA.axisData = {
        0, 0, 0,
        0, 0, 0
    };

    PositionData targetB;
    targetB.coordinateType = CoordinateType::Pulse;
    targetB.axisData = {
        0, 0, 0,
        -20000, -10000, -45000
    };

    PositionData robot_pos;
    ControllerStateData state;
    bool a = true;
    while (true) 
    {
        check(con->ControlGroup->ReadPositionData(ControlGroupId::R1,
            CoordinateType::BaseCoordinate, 0,
            tool, robot_pos),
            "Read robot pos");

        check(con->Status->ReadState(state), "Reading state");

        if (state.isRunning) {
            std::this_thread::sleep_for(std::chrono::milliseconds(100));
            std::cout << "waiting..." << std::endl;
            continue;
        }

        const auto& target = a ? targetA : targetB;
        a = !a;

        // I tried this together with the MotionStart below, but no change
        // check(con->MotionManager->MotionStop(), "extra stop");

        std::cout << "Sending movement:\n" << target << std::endl;
        LinearMotion motion(ControlGroupId::R1, target, 100, acc_dcc);
        check(con->MotionManager->AddPointToTrajectory(motion), "Add point");

        // works only the first time
        //check(con->MotionManager->MotionStart(), "Starting motion (2)");
        std::this_thread::sleep_for(std::chrono::milliseconds(100));
    }
}
```

**Reply by c68r69s** [reply]
Yes, I understand the check of running and the sleep is not required. This was mainly to not build an infinite queue, and to not get an overwhelming amount of debug output.

So I have copied your version of the code, and for me it still behaves the same way: The debug version works just fine (libYMConnect_D.so), the release version (libYMConnect.so) does nothing.
I did have to change the positions back to my original version in order to not collide with the security boundaries.

What I really don't get is the different behavior with the two versions of the library. I do not change anything in the test code or the compile options, I just link against libYMConnect.so or libYMConnect_D.so.

I also tried running it in valgrind, and while I see no errors, doing so causes none of the version to do anything. Could this be something timing related?

Thanks for your help

**Reply by ted-miller** [reply]
I see. I am now able to reproduce this same behavior. It doesn't appear to be a problem on your end. @aseamans and I will investigate.

**Comment by aseamans** [comment]
Just wanted to update you really quick. We have identified the problem and are working on a fix right now that we should be able to push out sometime this week. 
Thanks for your patience!

**Reply by c68r69s** [reply]
Sounds good, thanks for your help!

**Reply by c68r69s** [reply]
Hi - Any update on the issue? I'm sorry to bother, but this is part of our current project, and I need to decide on how to continue. So any update or ETA would be appreciated.

**Reply by aseamans** [reply, marked as answer]
Hi,
Sorry about the late reply and thank you for your patience.
We pushed out a new release that fixed the previous issue. 
Please let us know if it solves your issue, I verified that it runs with your code. 
Also let us know if any other issues arise.
Thank you!
-Austen

**Reply by c68r69s** [reply]
Hi, thanks a lot!  The original code now just works.
